kava/account.go

```diff
@@ -0,0 +1,207 @@
+type AccountBalanceService interface {
+	GetCoinsForSubAccount(
+		subAccount *types.SubAccountIdentifier,
+	) (sdk.Coins, error)
+}
+
+// BalanceServiceFactory provides an interface for creating a balance service for specifc a account and block
+type BalanceServiceFactory func(addr sdk.AccAddress, blockHeader *tmtypes.Header) (AccountBalanceService, error)
+
+// NewRPCBalanceFactory returns a balance service factory that uses an RPCClient to get an accounts balance
+func NewRPCBalanceFactory(rpc RPCClient) BalanceServiceFactory {
+	return func(addr sdk.AccAddress, blockHeader *tmtypes.Header) (AccountBalanceService, error) {
+		acc, err := rpc.Account(addr, blockHeader.Height)
+		if err != nil {
+			return nil, err
+		}
+
+		switch acc := acc.(type) {
+		case vestingexported.VestingAccount:
+			return &rpcVestingBalance{rpc: rpc, vacc: acc, blockHeader: blockHeader}, nil
+		default:
+			return &rpcBaseBalance{rpc: rpc, acc: acc, blockHeader: blockHeader}, nil
+		}
+	}
+}
+
+type rpcBaseBalance struct {
+	rpc         RPCClient
+	acc         authexported.Account
+	blockHeader *tmtypes.Header
+}
+
+func (b *rpcBaseBalance) GetCoinsForSubAccount(subAccount *types.SubAccountIdentifier) (coins sdk.Coins, err error) {
+	if subAccount == nil {
+		coins = b.acc.GetCoins()
+		return
+	}
+
+	switch subAccount.Address {
+	case AccLiquid:
+		coins = b.acc.SpendableCoins(b.blockHeader.Time)
+	case AccLiquidDelegated:
+		coins, err = b.totalDelegated()
+	case AccLiquidUnbonding:
+		coins, err = b.totalUnbondingDelegations()
+	default:
+		coins = sdk.Coins{}
+	}
+
+	return
+}
+
+func (b *rpcBaseBalance) totalDelegated() (sdk.Coins, error) {
+	delegations, err := b.rpc.Delegations(b.acc.GetAddress(), b.blockHeader.Height)
+	if err != nil {
+		return nil, err
+	}
+
+	return sumDelegations(delegations), nil
+}
+
+func (b *rpcBaseBalance) totalUnbondingDelegations() (sdk.Coins, error) {
+	unbondingDelegations, err := b.rpc.UnbondingDelegations(b.acc.GetAddress(), b.blockHeader.Height)
+	if err != nil {
+		return nil, err
+	}
+
+	return sumUnbondingDelegations(unbondingDelegations), nil
+}
+
+type rpcVestingBalance struct {
+	rpc         RPCClient
+	vacc        vestingexported.VestingAccount
+	blockHeader *tmtypes.Header
+}
+
+func (b *rpcVestingBalance) GetCoinsForSubAccount(subAccount *types.SubAccountIdentifier) (coins sdk.Coins, err error) {
+	if subAccount == nil {
+		coins = b.vacc.GetCoins()
+		return
+	}
+
+	switch subAccount.Address {
+	case AccLiquid:
+		coins = b.vacc.SpendableCoins(b.blockHeader.Time)
+	case AccVesting:
+		coins = b.vacc.GetVestingCoins(b.blockHeader.Time)
+	case AccLiquidDelegated:
+		coins, _, err = b.delegated()
+	case AccVestingDelegated:
+		_, coins, err = b.delegated()
+	case AccLiquidUnbonding:
+		coins, _, err = b.unbonding()
+	case AccVestingUnbonding:
+		_, coins, err = b.unbonding()
+	default:
+		coins = sdk.Coins{}
+	}
+
+	return
+}
+
+// delegated returns liquid and vesting coins that are staked
```

Comment:
Worth noting in the comment that this includes unbonding tokens

Reply:
I believe that's incorrect - Only staking tokens are included in the returned staked vesting and staked liquid.  Variable names could be improved in the function to reflect that.

I did find an edge case that isn't covered going back through the calculation, will adjust to make the method more clear and fix the edge case.

Reply:
Going to resolve edge case in another PR -- prefer to integrate this one first with other changes.